### PR TITLE
Reconcile semantics for Suggestion Algorithms

### LINKED ISSUE #1534: Grid Search stuck when parallelTrialCount < maxTrialCount
/kind bug

**What steps did you take and what happened:**

I created an experiment to perform a grid search, set `maxTrialCount` equal to the size of the grid, and set `parallelTrialCount < maxTrialCount`.

The first set of parallel trials completes successfully, but at some point after that the experiment gets stuck. The suggestion shows requested > assigned and has the following warning:
`Warning  ReconcileError  4s (x12 over 31s)  suggestion-controller  The response contains unexpected trials`

**What did you expect to happen:**
Experiment to execute trials for each point in the parameter space and then complete.

**Anything else you would like to add:**

I am running this image for the katib-controller: gcr.io/kubeflow-images-public/katib/v1alpha3/katib-controller:latest (I also tested with gcr.io/kubeflow-images-public/katib/v1alpha3/katib-controller:v0.8.0)

The issue seems like it is likely related to https://github.com/kubeflow/katib/issues/1494

If I set `parallelTrialCount >= maxTrialCount` (enabling all runs to execute immediately) the experiment succeeds.

**Environment:**

- Kubeflow version (`kfctl version`): v1.0.1
- Minikube version (`minikube version`): N/A
- Kubernetes version: (use `kubectl version`): v1.18.17-gke.100
- OS (e.g. from `/etc/os-release`):

---
experiment.yaml:
```
apiVersion: "kubeflow.org/v1alpha3"
kind: Experiment
metadata:
  namespace: sid
  labels:
    controller-tools.k8s.io: "1.0"
  name: minimal-grid
spec:
  objective:
    type: maximize
    goal: 999999
    objectiveMetricName: value
  resumePolicy: Never
  algorithm:
    algorithmName: grid
  parallelTrialCount: 5
  maxTrialCount: 15
  maxFailedTrialCount: 3
  parameters:
    - name: --param_1
      parameterType: int
      feasibleSpace:
        min: "0"
        max: "15"
  trialTemplate:
    goTemplate:
      rawTemplate: |-
        apiVersion: batch/v1
        kind: Job
        metadata:
          name: {{.Trial}}
          namespace: {{.NameSpace}}
        spec:
          template:
            spec:
              containers:
              - name: {{.Trial}}
                image: <IMAGE>
                command:
                - "python3"
                - "return_param_value.py"
                {{- with .HyperParameters}}
                {{- range .}}
                - "{{.Name}}={{.Value}}"
                {{- end}}
                {{- end}}
              restartPolicy: Never
```

Result of `kubectl describe suggestion <suggestion-name>`:
```
Name:         minimal-grid
Namespace:    sid
Labels:       controller-tools.k8s.io=1.0
Annotations:  <none>
API Version:  kubeflow.org/v1alpha3
Kind:         Suggestion
Metadata:
  Creation Timestamp:  2021-05-18T05:58:39Z
  Generation:          9
  Managed Fields:
    API Version:  kubeflow.org/v1alpha3
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
        f:labels:
          .:
          f:controller-tools.k8s.io:
        f:ownerReferences:
      f:spec:
        .:
        f:algorithmName:
        f:requests:
      f:status:
        .:
        f:conditions:
        f:startTime:
        f:suggestionCount:
        f:suggestions:
    Manager:    katib-controller
    Operation:  Update
    Time:       2021-05-18T05:58:56Z
  Owner References:
    API Version:           kubeflow.org/v1alpha3
    Block Owner Deletion:  true
    Controller:            true
    Kind:                  Experiment
    Name:                  minimal-grid
    UID:                   90e538b2-5858-45d2-af0e-38c0837d6a8d
  Resource Version:        69731263
  Self Link:               /apis/kubeflow.org/v1alpha3/namespaces/sid/suggestions/minimal-grid
  UID:                     3875d508-b3ee-4426-88a7-1b307c381249
Spec:
  Algorithm Name:  grid
  Requests:        13
Status:
  Conditions:
    Last Transition Time:  2021-05-18T05:58:39Z
    Last Update Time:      2021-05-18T05:58:39Z
    Message:               Suggestion is created
    Reason:                SuggestionCreated
    Status:                True
    Type:                  Created
    Last Transition Time:  2021-05-18T05:58:54Z
    Last Update Time:      2021-05-18T05:58:54Z
    Message:               Deployment is ready
    Reason:                DeploymentReady
    Status:                True
    Type:                  DeploymentReady
    Last Transition Time:  2021-05-18T05:58:55Z
    Last Update Time:      2021-05-18T05:58:55Z
    Message:               Suggestion is running
    Reason:                SuggestionRunning
    Status:                True
    Type:                  Running
  Start Time:              2021-05-18T05:58:39Z
  Suggestion Count:        8
  Suggestions:
    Name:  minimal-grid-dq75pd5j
    Parameter Assignments:
      Name:   --param_1
      Value:  0
    Name:     minimal-grid-jsl78k9b
    Parameter Assignments:
      Name:   --param_1
      Value:  1
    Name:     minimal-grid-nbhlss7f
    Parameter Assignments:
      Name:   --param_1
      Value:  2
    Name:     minimal-grid-p87q65dk
    Parameter Assignments:
      Name:   --param_1
      Value:  3
    Name:     minimal-grid-dt2s5jd5
    Parameter Assignments:
      Name:   --param_1
      Value:  4
    Name:     minimal-grid-jxtrdnk9
    Parameter Assignments:
      Name:   --param_1
      Value:  6
    Name:     minimal-grid-l2t2gpc8
    Parameter Assignments:
      Name:   --param_1
      Value:  9
    Name:     minimal-grid-4dpdlj2v
    Parameter Assignments:
      Name:   --param_1
      Value:  13
Events:
  Type     Reason          Age               From                   Message
  ----     ------          ----              ----                   -------
  Warning  ReconcileError  0s (x8 over 15s)  suggestion-controller  The response contains unexpected trials
```

### PR DESCRIPTION
**Pull request by johnugeorge**
Currently, GetSuggestions call does not follow Kubernetes reconcile semantics. eg: If suggestion controller cannot update the suggestions returned from GetSuggestions call(from suggestion algorithm service),  new suggestions are created again during the next try. This causes few suggestions to be leaked out. 

In this PR, new variable is passed in the GetSuggestions Call which indicates the total Suggestions requested till date. If there are more trials in DB which are not recorded, it reuses the missed suggestions from DB while remaining required number is generated. So, GetSuggestions will ensure that missed suggestions are reused first before generating new ones. 

Fixes #1534 
/hold

### CODE REVIEW

File: pkg/controller.v1beta1/suggestion/suggestionclient/suggestionclient.go [hunk 1]
```diff
@@ -102,21 +102,22 @@ func (g *General) SyncAssignments(
...
 
 	// Get new suggestions
 	responseSuggestion, err := rpcClientSuggestion.GetSuggestions(ctx, requestSuggestion)
 	if err != nil {
 		return err
 	}
-	logger.Info("Getting suggestions", "endpoint", endpoint, "response", responseSuggestion,
-		"request", requestSuggestion)
+	logger.Info("Getting suggestions", "endpoint", endpoint, "response", len(responseSuggestion.ParameterAssignments),
+		"requestNum", requestNum)
```
Comment (andreyvelich) [inline pkg/controller.v1beta1/suggestion/suggestionclient/suggestionclient.go:117]: Make it more clear here ?

```suggestion
	logger.Info("Getting suggestions", "endpoint", endpoint, "number of response parameters", len(responseSuggestion.ParameterAssignments),
		"number of request parameters", requestNum)
```

File: pkg/controller.v1beta1/suggestion/suggestionclient/suggestionclient.go [hunk 2]
```diff
@@ -102,21 +102,22 @@ func (g *General) SyncAssignments(
...
 	if len(responseSuggestion.ParameterAssignments) != requestNum {
 		err := fmt.Errorf("The response contains unexpected trials")
-		logger.Error(err, "The response contains unexpected trials", "requestNum", requestNum, "response", responseSuggestion)
+		logger.Error(err, "The response contains unexpected trials", "requestNum", requestNum, "response", len(responseSuggestion.ParameterAssignments))
```
Comment (andreyvelich) [inline pkg/controller.v1beta1/suggestion/suggestionclient/suggestionclient.go:120]:
```suggestion
		logger.Error(err, "The response contains unexpected trials", "number of request parameters", requestNum, "number of response parameters", len(responseSuggestion.ParameterAssignments))
```

File: pkg/mock/v1beta1/api/earlystopping.go [hunk 3]
```diff
@@ -6,10 +6,11 @@ package mock
 
 import (
 	context "context"
+	reflect "reflect"
```
Comment (andreyvelich) [inline pkg/mock/v1beta1/api/earlystopping.go:9]: We might need to define specific version for `mockgen` to have same generation files.
Do we need these changes in the PR ?

File: pkg/suggestion/v1beta1/chocolate/base_service.py [hunk 4]
```diff
@@ -157,9 +157,23 @@ def getSuggestions(self, trials, request_number):
                     logger.info("New record in sqlite DB is updated")
                     logger.info("{}\n".format(
                         self.created_trials[new_trial_loss_idx]))
+        # Assuming that created_trials are already populated
+        # TODO: Handle Restart of algorithm pod
+        logger.info("{} Trials created in DB".format(len(self.created_trials)))
+        new_actual_requested_no = total_request_number - len(self.created_trials)
+        prev_generated_no = request_number - new_actual_requested_no
+        logger.info("In this call, New {} Trials will be generated, {} Trials will be reused from previously generated".format(new_actual_requested_no, prev_generated_no))
```
Comment (andreyvelich) [inline pkg/suggestion/v1beta1/chocolate/base_service.py:165]: In the first call, `new_actual_requested_no` = 0 and `prev_generated_no` = 3, when `request_number` = 3.
Is that correct ?

File: pkg/suggestion/v1beta1/chocolate/base_service.py [hunk 5]
```diff
@@ -157,9 +157,23 @@ def getSuggestions(self, trials, request_number):
...
+        if total_request_number != len(self.created_trials) + request_number:
+            logger.info("Mismatch in generated trials with k8s suggestions trials")
```
Comment (andreyvelich) [inline pkg/suggestion/v1beta1/chocolate/base_service.py:167]: What does this log mean ?

File: pkg/suggestion/v1beta1/chocolate/base_service.py (hunk 4 above)
Comment (johnugeorge) [reply pkg/suggestion/v1beta1/chocolate/base_service.py:165]: It is the other way around. 
In the normal case,  
         total_request_number == len(self.created_trials) + request_number where self.created_trials correspond to the number of previously created trials.  So, prev_generated_no will be 0 in this case 

When there is a difference, it means that some of the suggestions in self.created_trials(same as in DB)  are not recorded in K8s Suggestions resource. So,  prev_generated_no will be greater than 0 in this case

File: pkg/suggestion/v1beta1/chocolate/base_service.py (hunk 5 above)
Comment (johnugeorge) [reply pkg/suggestion/v1beta1/chocolate/base_service.py:167]: Earlier comment

**Review by andreyvelich — COMMENTED** [review]
I tested this fix.

/lgtm
/approve
/hold cancel

**Pull request merged**